### PR TITLE
Locale independent URL parsing

### PR DESCRIPTION
**Pull request by GrahamCampbell**
Co-Authored-By: Alexey Shokov <alexey@shockov.com>. Replaces #297.

### CODE REVIEW

**Review by Nyholm — APPROVED** [review]
Thank you

**Comment by albe** [comment]
This change seems to be a breaking change with IPv6 addresses.

`parse_url('http://[2a00:f48:1008::212:183:10]')` correctly returns `'[2a00:f48:1008::212:183:10]'` for the host part. This change instead returns `[2a00:f48:1008::212:183`, which is wrong and breaks URLs.

**Comment by Nyholm** [comment]
Thank you. 
I'll check

**Comment by albe** [comment]
I added a test for this in https://github.com/guzzle/psr7/pull/401

File: src/Uri.php [hunk 1]
```diff
@@ -582,7 +616,7 @@ private function filterScheme($scheme)
             throw new \InvalidArgumentException('Scheme must be a string');
         }
 
-        return strtolower($scheme);
+        return \strtr($scheme, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz');
```
Comment (Tobion) [inline src/Uri.php:619]: This is irrelevant with https://wiki.php.net/rfc/strtolower-ascii and I don't think we should add workarounds for things that even PHP core changes and think is stupid and not meant to be used.
Also there isn't and test for this added. So I'm 👎

Comment (GrahamCampbell) [reply src/Uri.php:619]: I guess we can re-visit this in a few years, when we drop PHP less than 8.2.

**Pull request merged**